yjit/src/backend/ir.rs

```diff
@@ -1646,9 +1646,19 @@ impl Assembler {
     }
 
     pub fn ccall(&mut self, fptr: *const u8, opnds: Vec<Opnd>) -> Opnd {
-        assert_eq!(self.ctx.get_reg_temps(), RegTemps::default(), "temps must be spilled before ccall");
+        // Save what registers are available for the C arguments.
+        let reg_temps = self.ctx.get_reg_temps();
+        // Spill stack temp registers since they are caller-saved registers.
+        // Note that this doesn't spill stack temps that are already popped
+        // but may still be used in the C arguments.
+        self.spill_temps();
+
         let out = self.next_opnd_out(Opnd::match_num_bits(&opnds));
+        // For C arguments, use stack temp registers that are still alive.
+        // After the C call that may clobber them, use spilled stack temps.
+        self.ctx.set_reg_temps(reg_temps);
         self.push_insn(Insn::CCall { fptr, opnds, out });
+        self.ctx.set_reg_temps(RegTemps::default());
```

Comment:
Trying to make sure I understand what this PR does.

From the PR title it sounded like you are passing arguments via registers instead of through the temp stack, but looking at the code, it looks like the PR is doing something different.

You've moved the call to `spill_temps()` inside of `ccall()`, so that you can be smarter about which temps you spill instead of always spilling, correct?

I don't quite understand what the two calls to `set_reg_temps()` before and after the `Insn::CCall` are doing. I feel like the comment there could potentially be improved.

Reply:
Essentially, this PR achieves a couple of optimizations:

1. Skip spilling off-stack `Opnd::Stack`s on `ccall()`. They'll also be passed to C arguments as registers.
2. Pass on-stack `Opnd::Stack`s to C arguments as registers _after spilling them_ on `ccall()`.

> You've moved the call to spill_temps() inside of ccall(), so that you can be smarter about which temps you spill instead of always spilling, correct?

So this discusses (1), but you may be missing (2).

Let's say we have this stack when entering an insn, `[A opnd (spilled), B opnd (register), C opnd (register)]`, and `C opnd` is popped before a `ccall()`. All of the operands are used by the call.

With (1), it no longer spills `C opnd`. `C opnd` is passed as a register as well.

With (2), it first spills `B opnd`. But at that moment, `B opnd` is also in a register. So we use a register to pass `B opnd`, which wasn't the case before this PR.

The resulting code looks like:

* before
    1. spill B
    2. spill C -- removed by (1)
    3. pass A (spilled)
    4. pass B (spilled) -- optimized by (2)
    5. pass C (spilled) -- optimized by (1)
* after
    1. spill B
    2. pass A (spilled)
    3. pass B (register)
    4. pass C (register)

Reply:
> I feel like the comment there could potentially be improved.

How about this? https://github.com/ruby/ruby/pull/7920/commits/bcce57a4f5b2b8d436044aa94dad0e2c1c372464